### PR TITLE
Added a timeline and a progress bar

### PR DESCRIPTION
**Pull request by Forbaya**
This is the result of our two last sprints. We added a timeline page and a progress bar to the exercise submission dialog.

### CODE REVIEW

File: tmc-plugin/src/fi/helsinki/cs/tmc/model/CourseDb.java [hunk 1]
```diff
@@ -85,25 +87,53 @@ public void setCurrentCourseName(String currentCourseName) {
...
+        for (int i = 1; i <= getCurrentCourseWeekCount(); i++) {
+            boolean wholeWeekDone = true;
+            for (Exercise ex : getExercisesByWeek(i)) {
+                if (!ex.isCompleted()) {
+                    wholeWeekDone = false;
+                    break;
+                }
+            }
+            if (wholeWeekDone) {
+                week = i;
```
Comment (Piia) [inline tmc-plugin/src/fi/helsinki/cs/tmc/model/CourseDb.java]: This method is for finding the latest fully completed week. It should iterate through weeks and refresh to-be-returned variable "week" every time the week is fully completed. We haven't been able to test it yet.

Comment (Piia) [inline tmc-plugin/src/fi/helsinki/cs/tmc/model/CourseDb.java]: Adaptive exercises are available only after a fully completed week. This method shouldn't mind if a user left one week undone. We haven't really had the opportunity to test it though... :(

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 2]
```diff
@@ -1,72 +1,72 @@
...
+                });
+
+
+                // I hope the above stuff will work so that
+                // we don't need this commented fix anymore
+//                if (courseDb.getCurrentCourseExercises().contains(ex)) {
+//                    courseDb.getCurrentCourseExercises().remove(ex);
+//                }
+//                courseDb.getCurrentCourseExercises().add(ex);
+//                courseDb.exerciseDownloaded(ex);
```
Comment (Piia) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: Yes, it works. Deleting...

File: tmc-plugin/src/fi/helsinki/cs/tmc/timeline/TimelineJson.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        TreeMap<String, String> exerciseMap = new TreeMap<>();
+        for (Exercise ex : exercises) {
+            if (!ex.isAdaptive()) {
+                String color = getExerciseColor(ex);
+                exerciseMap.put(ex.getName(), color);
+            }
+        }
+        Gson gson = new GsonBuilder().create();
+        String json = gson.toJson(exerciseMap);
+        logger.log(Level.INFO, json);
```
Comment (Piia) [inline tmc-plugin/src/fi/helsinki/cs/tmc/timeline/TimelineJson.java]: Yes. Now that it works, we don't need that information anymore. Deleting...

File: tmc-plugin/src/fi/helsinki/cs/tmc/ui/SuccessfulSubmissionDialog.java [hunk 4]
```diff
@@ -25,29 +27,26 @@
...
         this.setTitle(exercise.getName() + " passed");
-        
+
         JPanel contentPane = new JPanel();
         contentPane.setBorder(BorderFactory.createEmptyBorder(8, 8, 8, 8));
         contentPane.setLayout(new BoxLayout(contentPane, BoxLayout.Y_AXIS));
         setContentPane(contentPane);
-        
-        addYayLabel();
+        //setAlwaysOnTop(true);
```
Comment (Redande) [inline tmc-plugin/src/fi/helsinki/cs/tmc/ui/SuccessfulSubmissionDialog.java]: Dead code

File: tmc-plugin/src/fi/helsinki/cs/tmc/ui/SuccessfulSubmissionDialog.java [hunk 5]
```diff
@@ -179,5 +180,10 @@ public void actionPerformed(ActionEvent e) {
         });
         getContentPane().add(hbox(hglue(), okButton));
     }
-    
+
+    private void addProgressPanel(Exercise exercise) {
+        JPanel progressPanel = new ProgressPanel(exercise);
+        //progressPanel.setAlignmentX(CENTER_ALIGNMENT);
```
Comment (Redande) [inline tmc-plugin/src/fi/helsinki/cs/tmc/ui/SuccessfulSubmissionDialog.java]: Dead code

File: tmc-plugin/src/fi/helsinki/cs/tmc/model/CourseDb.java (hunk 1 above)
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/model/CourseDb.java]: Taking a longer and less coffee-starved look at the logic here, this seems fine.

Comment (Piia) [inline tmc-plugin/src/fi/helsinki/cs/tmc/model/CourseDb.java]: ok :D

**Pull request merged**